src/VisualStudio/Core/Def/LanguageService/AbstractLanguageService`2.VsCodeWindowManager.cs

```diff
@@ -224,6 +232,66 @@ public int RemoveAdornments()
 
                 return VSConstants.S_OK;
             }
+
+            // We expect GetOutline to be called every time a new code window is created.
+            int IVsDocOutlineProvider.GetOutline(out IntPtr phwnd, out IOleCommandTarget? ppCmdTarget)
+            {
+                var languageServiceBroker = _languageService.Package.ComponentModel.GetService<ILanguageServiceBrokerShim>();
+                var threadingContext = _languageService.Package.ComponentModel.GetService<IThreadingContext>();
+                var asyncListenerProvider = _languageService.Package.ComponentModel.GetService<IAsynchronousOperationListenerProvider>();
+                var asyncListener = asyncListenerProvider.GetListener(FeatureAttribute.DocumentOutline);
+                var editorAdaptersFactoryService = _languageService.Package.ComponentModel.GetService<IVsEditorAdaptersFactoryService>();
+                _documentOutlineCts = new CancellationTokenSource();
+
+                threadingContext.ThrowIfNotOnUIThread();
+
+                // Overwrite the existing host with a new Document Outline for this code window.
+                _documentOutlineViewHost = new ElementHost
+                {
+                    Dock = DockStyle.Fill,
+                    Child = new DocumentOutlineControl(
+                        languageServiceBroker, threadingContext, asyncListener, editorAdaptersFactoryService, _codeWindow, _documentOutlineCts.Token)
+                };
+
+                phwnd = _documentOutlineViewHost.Handle;
+                ppCmdTarget = null;
+
+                Logger.Log(FunctionId.DocumentOutline_WindowOpen);
+
+                return VSConstants.S_OK;
+            }
+
+            int IVsDocOutlineProvider.ReleaseOutline(IntPtr hwnd, IOleCommandTarget pCmdTarget)
+            {
+                var threadingContext = _languageService.Package.ComponentModel.GetService<IThreadingContext>();
+                threadingContext.ThrowIfNotOnUIThread();
+
+                _documentOutlineViewHost?.Dispose();
```

Comment:
Does nothing dispose/disconnect the underlying DocumentOutlineControl as well? I guess I would have imagined we needed to do that to ensure we've disconnected from any other event handlers, but maybe not? Does this run if the document outline window is closed explicitly by the user?